### PR TITLE
Purge left nodes along with dead ones (#253)

### PR DESCRIPTION
**Pull request by bwaters**
Keep the state from infinitely expanding

### CODE REVIEW

File: util.go [hunk 1]
```diff
@@ -96,13 +96,13 @@ func pushPullScale(interval time.Duration, n int) time.Duration {
...
 
-// moveDeadNodes moves nodes that are dead and beyond the gossip to the dead interval
+// moveDeadNodes moves nodes that are dead and left and beyond the gossip to the dead interval
 // to the end of the slice and returns the index of the first moved node.
 func moveDeadNodes(nodes []*nodeState, gossipToTheDeadTime time.Duration) int {
 	numDead := 0
 	n := len(nodes)
 	for i := 0; i < n-numDead; i++ {
-		if nodes[i].State != StateDead {
+		if nodes[i].State != StateDead && nodes[i].State != StateLeft {
```
Comment (i0rek) [inline util.go:105]:
```suggestion
		if !nodes[i].DeadOrLeft() {
```

Comment (i0rek) [reply util.go:105]: This is a bit easier to read in my mind and aligns with the rest of the code.

**Review by i0rek — CHANGES_REQUESTED** [review]
This looks good to me, thanks for the detailed explanation in the corresponding issue. Could you modify `TestMoveDeadNodes` to have a nodestate in `StateDead` to make sure it is working as intended?

File: util.go [hunk 2]
```diff
@@ -96,13 +96,13 @@ func pushPullScale(interval time.Duration, n int) time.Duration {
 	return time.Duration(multiplier) * interval
 }
 
-// moveDeadNodes moves nodes that are dead and beyond the gossip to the dead interval
+// moveDeadNodes moves nodes that are dead and left and beyond the gossip to the dead interval
```
Comment (dnephin) [inline util.go:99]: I'm not sure I follow this godoc comment, is there an extra word or a missing word here?

Does this sound right ?

```suggestion
// moveDeadNodes moves dead and left nodes that that have not changed during the gossipToTheDeadTime duration
```

**Review by dnephin — COMMENTED** [review]
Thanks for the fix!  One suggestion for the godoc.

Is there any way we can test this from within memberlist? If a full integration test makes it difficult to reproduce, maybe just a test of `moveDeadNodes`, to show that it moves nodes in both states? Edit: Oh, I missed that Hans already suggested a test for this, that sounds good!

File: util.go (hunk 2 above)
Comment (markan) [reply util.go:99]: As far as a test goes, maybe something in state_test:TestMemberList_ResetNodes?

File: util_test.go [hunk 3]
```diff
@@ -204,14 +218,19 @@ func TestMoveDeadNodes(t *testing.T) {
 			if nodes[i].State != StateDead {
 				t.Fatalf("Bad state %d", i)
 			}
+		case 3:
+			//Recently left node should remain at 3
+			if nodes[i].State != StateLeft {
+				t.Fatalf("Bad State #{i}")
```
Comment (kisunji) [inline util_test.go:224]:
```suggestion
				t.Fatalf("Bad State %d", i)
```

Comment (bwaters) [reply util_test.go:224]: If we are changing it here we should change it in the rest of the test.

Comment (kisunji) [reply util_test.go:224]: Perhaps you are using Goland with the default code folding? It compacts format strings like
```go
t.Fatalf("Bad State %d", i)
```
into
```go
t.Fatalf("Bad State #{i}")
```
But the latter is not a valid format syntax

Comment (bwaters) [reply util_test.go]: yes you were right ide was playing tricks on me

**Review by i0rek — APPROVED** [review]
LGTM

**Pull request merged**